File: packages/server/src/plugins/voice-mode.ts
import type { Logger } from "../logger"
import type { ClientConnectionManager, ClientConnectionRef } from "../clients/connection-manager"
import type { PluginChannelManager } from "./channel"

interface VoiceModeManagerOptions {
  connections: ClientConnectionManager
  channel: PluginChannelManager
  logger: Logger
}

export class VoiceModeManager {
  private readonly enabledConnectionsByInstance = new Map<string, Set<string>>()
  private readonly aggregateByInstance = new Map<string, boolean>()

  constructor(private readonly options: VoiceModeManagerOptions) {
    this.options.connections.subscribe((event) => {
      if (event.type !== "disconnected") return
      this.clearConnection(event.connection)
    })
  }

  setEnabled(instanceId: string, connection: ClientConnectionRef, enabled: boolean): boolean {
    if (enabled && !this.options.connections.isConnected(connection)) {
      this.options.logger.debug(
        { instanceId, clientId: connection.clientId, connectionId: connection.connectionId },
        "Ignoring voice mode enable for disconnected client connection",
      )
      return false
    }

    const key = getConnectionKey(connection)
    const current = this.enabledConnectionsByInstance.get(instanceId) ?? new Set<string>()

    if (enabled) {
      current.add(key)
      this.enabledConnectionsByInstance.set(instanceId, current)
    } else if (current.delete(key)) {
      if (current.size === 0) {
        this.enabledConnectionsByInstance.delete(instanceId)
      } else {
        this.enabledConnectionsByInstance.set(instanceId, current)
      }
    }

    this.options.logger.debug({ instanceId, clientId: connection.clientId, connectionId: connection.connectionId, enabled }, "Voice mode updated for client connection")
    this.publishIfChanged(instanceId)
    return true
  }

  syncInstance(instanceId: string): void {
    this.options.channel.send(instanceId, buildVoiceModeEvent(this.isEnabled(instanceId)))
  }

  isEnabled(instanceId: string): boolean {
    return this.aggregateByInstance.get(instanceId) === true
  }

  private clearConnection(connection: ClientConnectionRef): void {
    const key = getConnectionKey(connection)
    for (const [instanceId, enabledConnections] of Array.from(this.enabledConnectionsByInstance.entries())) {
      if (!enabledConnections.delete(key)) continue
      if (enabledConnections.size === 0) {
        this.enabledConnectionsByInstance.delete(instanceId)
      }
      this.publishIfChanged(instanceId)
    }
  }

  private publishIfChanged(instanceId: string): void {
    const enabled = (this.enabledConnectionsByInstance.get(instanceId)?.size ?? 0) > 0
    const previous = this.aggregateByInstance.get(instanceId) === true
    if (enabled === previous) return

    if (enabled) {
      this.aggregateByInstance.set(instanceId, true)
    } else {
      this.aggregateByInstance.delete(instanceId)
    }

    this.options.logger.debug(
      { instanceId, enabled },
      "Broadcasting aggregate voice mode",
    )
    this.options.channel.send(instanceId, buildVoiceModeEvent(enabled))
  }
}

function buildVoiceModeEvent(enabled: boolean) {
  return {
    type: "embeddedcowork.voiceMode",
    properties: {
      enabled,
      formatVersion: "v1",
    },
  }
}

function getConnectionKey(connection: ClientConnectionRef): string {
  return `${connection.clientId}:${connection.connectionId}`
}

File: packages/server-rust/src/plugins/voice_mode.rs
use std::collections::{HashMap, HashSet};
use std::sync::Arc;
use tokio::sync::Mutex;

use crate::clients::connection_manager::{ClientConnectionRef, ClientConnectionManager};
use crate::logger::Logger;
use crate::plugins::channel::{PluginChannelManager, PluginOutboundEvent};

pub struct VoiceModeManager {
    enabled_connections_by_instance: HashMap<String, HashSet<String>>,
    aggregate_by_instance: HashMap<String, bool>,
    connections: Arc<Mutex<ClientConnectionManager>>,
    channel: Arc<Mutex<PluginChannelManager>>,
    logger: Logger,
}

impl VoiceModeManager {
    pub fn new(
        connections: Arc<Mutex<ClientConnectionManager>>,
        channel: Arc<Mutex<PluginChannelManager>>,
        logger: Logger,
    ) -> Self {
        Self {
            enabled_connections_by_instance: HashMap::new(),
            aggregate_by_instance: HashMap::new(),
            connections,
            channel,
            logger,
        }
    }

    pub async fn set_enabled(
        &mut self,
        instance_id: &str,
        connection: &ClientConnectionRef,
        enabled: bool,
    ) -> bool {
        if enabled && !self.connections.lock().await.is_connected(connection) {
            tracing::debug!(
                component = %self.logger.component,
                instance_id = %instance_id,
                "Ignoring voice mode enable for disconnected client"
            );
            return false;
        }

        let key = get_connection_key(connection);
        let current = self.enabled_connections_by_instance
            .entry(instance_id.to_string())
            .or_insert_with(HashSet::new);

        if enabled {
            current.insert(key);
        } else {
            current.remove(&key);
            if current.is_empty() {
                self.enabled_connections_by_instance.remove(instance_id);
            }
        }

        self.publish_if_changed(instance_id).await;
        true
    }

    pub async fn sync_instance(&self, instance_id: &str) {
        let enabled = self.is_enabled(instance_id);
        let event = build_voice_mode_event(enabled);
        self.channel.lock().await.send(instance_id, &event);
    }

    pub fn is_enabled(&self, instance_id: &str) -> bool {
        self.aggregate_by_instance.get(instance_id) == Some(&true)
    }

    async fn publish_if_changed(&mut self, instance_id: &str) {
        let enabled = self.enabled_connections_by_instance
            .get(instance_id)
            .map(|s| !s.is_empty())
            .unwrap_or(false);

        let previous = self.aggregate_by_instance.get(instance_id) == Some(&true);
        if enabled == previous {
            return;
        }

        if enabled {
            self.aggregate_by_instance.insert(instance_id.to_string(), true);
        } else {
            self.aggregate_by_instance.remove(instance_id);
        }

        let event = build_voice_mode_event(enabled);
        self.channel.lock().await.send(instance_id, &event);
    }
}

fn build_voice_mode_event(enabled: bool) -> PluginOutboundEvent {
    let mut properties = std::collections::HashMap::new();
    properties.insert("enabled".to_string(), serde_json::Value::Bool(enabled));
    properties.insert("formatVersion".to_string(), serde_json::Value::String("v1".to_string()));

    PluginOutboundEvent {
        event_type: "embeddedcowork.voiceMode".to_string(),
        properties: Some(properties),
    }
}

fn get_connection_key(connection: &ClientConnectionRef) -> String {
    format!("{}:{}", connection.client_id, connection.connection_id)
}
